clinica/pydra/tasks/bids/utils.py

```diff
@@ -0,0 +1,79 @@
+from typing import Optional
+
+from pydra.engine.task import FunctionTask
+
+DEFAULT_QUERY = {
+    "bold": {
+        "suffix": "bold",
+        "extension": [".nii", ".nii.gz"],
+    },
+    "T1w": {
+        "suffix": "T1w",
+        "extension": [".nii", ".nii.gz"],
+    },
+}
+
+
+def read_bids(output_query: Optional[dict] = None, **kwargs) -> FunctionTask:
+    """Generate a BIDS reading task.
+
+    Parameters
+    ----------
+    output_query : dict
+        Mapping between output and BIDS query
+    **kwargs
+        Additional arguments passed to the task constructor
+
+    Returns
+    -------
+    pydra.engine.task.FunctionTask
+        BIDS reading task
+
+    Examples
+    --------
+    >>> task = read_bids()
+    >>> task.output_names
+    ['bold', 'T1w']
+    """
+    from os import PathLike
+
+    from pydra.engine.specs import BaseSpec, SpecInfo
+
+    output_query = output_query or DEFAULT_QUERY
+
+    def inner(base_dir):
+        from ancpbids import BIDSLayout
+
+        layout = BIDSLayout(ds_dir=base_dir)
+
+        results = [
+            layout.get(return_type="files", **query)
+            for key, query in list(output_query.items())
+        ]
+
+        return tuple(results) if len(results) > 1 else results[0]
```

Comment:
I don't understand why the function is changing the return type here for the two use cases?

Also  It seems that when I try to introduce another entity to be read (I tried with bold) the N4BiasFieldCorrectionInputSpec does not receive the correct type (it receives a list. Granted, this second problem was probably present in the previous implementation as well but it's worthwhile thinking about if we are to change the return type.

Reply:
> I don't understand why the function is changing the return type here for the two use cases?

There is a difference between returning a tuple composed of a single list (such as -> Tuple[list[str]]) and a single list (-> list[str]). Pydra seems sensitive to that difference.

> N4BiasFieldCorrectionInputSpec does not receive the correct type (it receives a list.

You were right to test this. Sounds more like an issue with the splitter, not the reader.

Reply:
> Pydra seems sensitive to that difference.

Probably due to this [section](https://github.com/nipype/pydra/blob/ac06f8cc691c62e3d56e9e94ef015a675c73851f/pydra/engine/task.py#L196) of the code.

Reply:
> Can you illustrate with one use case where a tuple is needed?

You just tried a query with both `T1w` and `bold`. What do you think should be returned?

`-> Tuple[list(str), list(str)]`, one for the list of T1w files, one for the list of bold files.

Reply:
> > I think it makes sense to add the functionality to filter on subject/session in this PR.
> 
> That's gonna take longer then. Is that ok?

I mean it's primary reason brought forward motivating the use of `ancpbids` so it seems  the PR is not complete without this feature.

Reply:
> What do you think should be returned?

I'm not sure what should be returned but I'm asking why you chose a tuple over a list (of lists) for instance.

Reply:
That's because if you want N named outputs (for instance `T1w` and `bold`), you need to return a tuple of N elements. Here is the relevant part of the [code](https://github.com/nipype/pydra/blob/ac06f8cc691c62e3d56e9e94ef015a675c73851f/pydra/engine/task.py#L199) for reference.

Returning a `list` or `list(list)` will yield only a single output, or fail with a `RuntimeError` if the shape of the returned type and the list of `output_names` don't match.

Reply:
> That's because if you want N named outputs (for instance `T1w` and `bold`), you need to return a tuple of N elements. Here is the relevant part of the [code](https://github.com/nipype/pydra/blob/ac06f8cc691c62e3d56e9e94ef015a675c73851f/pydra/engine/task.py#L199) for reference.
> 
> Returning a `list` or `list(list)` will yield only a single output, or fail with a `RuntimeError` if the shape of the returned type and the list of `output_names` don't match.

Clear. Thanks for the reference.